### ISSUE TITLE
Not finding cuda during execution

### ISSUE THREAD

**Issue by lennart-maack** [huggingface/competitions#47]
For my competition, I want participants of the challenge to have access to a GPU. Therefore, I selected the Nvidia T4 small as space Hardware.
In the conf.json, I sett `"HARDWARE": "t4-small"`
My requirements.txt looks like this:

```
--extra-index-url https://download.pytorch.org/whl/cu113
 torch
 torchvision
 pycocotools
 pandas
 timm
 pillow
 transformers
 tqdm
```

When making a new submission, I can not access GPU somehow.
When running `torch.cuda.is_available()` inside script.py, I get False as a return.

My logs in the submission container look like this:

===== Application Startup at 2024-10-25 13:11:14 =====

==========
== CUDA ==
==========

CUDA Version 12.1.1

Container image Copyright (c) 2016-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

This container image and its contents are governed by the NVIDIA Deep Learning Container License.
By pulling and using the container, you accept the terms and conditions of this license:
https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license

A copy of this license is made available in this container at /NGC-DL-CONTAINER-LICENSE for your convenience.

INFO:     Started server process [34]
INFO:     Waiting for application startup.
2024-10-25 13:14:36.258 | INFO     | competitions.utils:run_evaluation:160 - ['python', '-m', 'competitions.evaluate', '--config', '/tmp/model/params.json']
2024-10-25 13:14:36.265 | INFO     | competitions.api:lifespan:93 - Started training with PID 36
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:7860 (Press CTRL+C to quit)
2024-10-25 13:14:38.329 | INFO     | __main__:run:93 - competition_id='Mtec-Tuhh/ISMChallengePhase2b' competition_type='script' metric='custom' token='hf_******' team_id='40873f62-50e5-4cc1-b54b-f3fe66f4c8d6' submission_id='9f68f5d4350139cd06e1e52ef8521c5e627dd1b3__6db9c96e-4f92-4b9b-a08d-99c06a745728' submission_id_col='file_name' submission_cols=['file_name', 'bbox', 'category_id'] submission_rows=10000 output_path='/tmp/model' submission_repo='LennartMaack/ISMChallengePhase2bSubmission' time_limit=3600 dataset='LennartMaack/ISMChallengePhase2a_private' submission_filenames=['submission.csv']
2024-10-25 13:14:38.807 | INFO     | __main__:run:111 - Installing requirements
ERROR: xtra-index-url is not a valid editable requirement. It should either be a path to a local project or a VCS URL (beginning with bzr+http, bzr+https, bzr+ssh, bzr+sftp, bzr+ftp, bzr+lp, bzr+file, git+http, git+https, git+ssh, git+git, git+file, hg+file, hg+http, hg+https, hg+ssh, hg+static-http, svn+ssh, svn+http, svn+https, svn+svn, svn+file).
2024-10-25 13:14:39.254 | INFO     | competitions.utils:uninstall_requirements:273 - Requirements uninstalled.
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cu113
Requirement already satisfied: torch in ./env/lib/python3.10/site-packages (from -r install.txt (line 2)) (2.5.0)
Requirement already satisfied: torchvision in ./env/lib/python3.10/site-packages (from -r install.txt (line 3)) (0.20.0)
Collecting pycocotools (from -r install.txt (line 4))
  Downloading pycocotools-2.0.8-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.1 kB)
Requirement already satisfied: pandas in ./env/lib/python3.10/site-packages (from -r install.txt (line 5)) (2.2.2)
Collecting timm (from -r install.txt (line 6))
  Downloading timm-1.0.11-py3-none-any.whl.metadata (48 kB)
Requirement already satisfied: pillow in ./env/lib/python3.10/site-packages (from -r install.txt (line 7)) (10.4.0)
Requirement already satisfied: transformers in ./env/lib/python3.10/site-packages (from -r install.txt (line 8)) (4.44.1)
Requirement already satisfied: tqdm in ./env/lib/python3.10/site-packages (from -r install.txt (line 9)) (4.66.4)
Requirement already satisfied: filelock in ./env/lib/python3.10/site-packages (from torch->-r install.txt (line 2)) (3.13.1)
Requirement already satisfied: typing-extensions>=4.8.0 in ./env/lib/python3.10/site-packages (from torch->-r install.txt (line 2)) (4.11.0)
Requirement already satisfied: networkx in ./env/lib/python3.10/site-packages (from torch->-r install.txt (line 2)) (3.2.1)
Requirement already satisfied: jinja2 in ./env/lib/python3.10/site-packages (from torch->-r install.txt (line 2)) (3.1.4)
Requirement already satisfied: fsspec in ./env/lib/python3.10/site-packages (from torch->-r install.txt (line 2)) (2024.6.1)
Requirement already satisfied: sympy==1.13.1 in ./env/lib/python3.10/site-packages (from torch->-r install.txt (line 2)) (1.13.1)
Requirement already satisfied: mpmath<1.4,>=1.1.0 in ./env/lib/python3.10/site-packages (from sympy==1.13.1->torch->-r install.txt (line 2)) (1.3.0)
Requirement already satisfied: numpy in ./env/lib/python3.10/site-packages (from torchvision->-r install.txt (line 3)) (1.26.4)
Requirement already satisfied: matplotlib>=2.1.0 in ./env/lib/python3.10/site-packages (from pycocotools->-r install.txt (line 4)) (3.9.2)
Requirement already satisfied: python-dateutil>=2.8.2 in ./env/lib/python3.10/site-packages (from pandas->-r install.txt (line 5)) (2.9.0.post0)
Requirement already satisfied: pytz>=2020.1 in ./env/lib/python3.10/site-packages (from pandas->-r install.txt (line 5)) (2024.2)
Requirement already satisfied: tzdata>=2022.7 in ./env/lib/python3.10/site-packages (from pandas->-r install.txt (line 5)) (2024.2)
Requirement already satisfied: pyyaml in ./env/lib/python3.10/site-packages (from timm->-r install.txt (line 6)) (6.0.2)
Requirement already satisfied: huggingface_hub in ./env/lib/python3.10/site-packages (from timm->-r install.txt (line 6)) (0.24.6)
Requirement already satisfied: safetensors in ./env/lib/python3.10/site-packages (from timm->-r install.txt (line 6)) (0.4.5)
Requirement already satisfied: packaging>=20.0 in ./env/lib/python3.10/site-packages (from transformers->-r install.txt (line 8)) (24.1)
Requirement already satisfied: regex!=2019.12.17 in ./env/lib/python3.10/site-packages (from transformers->-r install.txt (line 8)) (2024.9.11)
Requirement already satisfied: requests in ./env/lib/python3.10/site-packages (from transformers->-r install.txt (line 8)) (2.32.3)
Requirement already satisfied: tokenizers<0.20,>=0.19 in ./env/lib/python3.10/site-packages (from transformers->-r install.txt (line 8)) (0.19.1)
Requirement already satisfied: contourpy>=1.0.1 in ./env/lib/python3.10/site-packages (from matplotlib>=2.1.0->pycocotools->-r install.txt (line 4)) (1.3.0)
Requirement already satisfied: cycler>=0.10 in ./env/lib/python3.10/site-packages (from matplotlib>=2.1.0->pycocotools->-r install.txt (line 4)) (0.12.1)
Requirement already satisfied: fonttools>=4.22.0 in ./env/lib/python3.10/site-packages (from matplotlib>=2.1.0->pycocotools->-r install.txt (line 4)) (4.54.1)
Requirement already satisfied: kiwisolver>=1.3.1 in ./env/lib/python3.10/site-packages (from matplotlib>=2.1.0->pycocotools->-r install.txt (line 4)) (1.4.7)
Requirement already satisfied: pyparsing>=2.3.1 in ./env/lib/python3.10/site-packages (from matplotlib>=2.1.0->pycocotools->-r install.txt (line 4)) (3.2.0)
Requirement already satisfied: six>=1.5 in ./env/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas->-r install.txt (line 5)) (1.16.0)
Requirement already satisfied: MarkupSafe>=2.0 in ./env/lib/python3.10/site-packages (from jinja2->torch->-r install.txt (line 2)) (2.1.3)
Requirement already satisfied: charset-normalizer<4,>=2 in ./env/lib/python3.10/site-packages (from requests->transformers->-r install.txt (line 8)) (3.3.2)
Requirement already satisfied: idna<4,>=2.5 in ./env/lib/python3.10/site-packages (from requests->transformers->-r install.txt (line 8)) (3.7)
Requirement already satisfied: urllib3<3,>=1.21.1 in ./env/lib/python3.10/site-packages (from requests->transformers->-r install.txt (line 8)) (2.2.3)
Requirement already satisfied: certifi>=2017.4.17 in ./env/lib/python3.10/site-packages (from requests->transformers->-r install.txt (line 8)) (2024.8.30)
Downloading pycocotools-2.0.8-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (427 kB)
Downloading timm-1.0.11-py3-none-any.whl (2.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 48.5 MB/s eta 0:00:00
Installing collected packages: pycocotools, timm
Successfully installed pycocotools-2.0.8 timm-1.0.11
2024-10-25 13:14:41.783 | INFO     | competitions.utils:install_requirements:304 - Requirements installed.
2024-10-25 13:16:12.987 | INFO     | __main__:generate_submission_file:30 - Downloading submission dataset
2024-10-25 13:16:16.806 | INFO     | __main__:generate_submission_file:41 - Generating submission file
The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.
0it [00:00, ?it/s]
0it [00:00, ?it/s]

Fri Oct 25 13:16:20 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.129.03             Driver Version: 535.129.03   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       On  | 00000000:00:1E.0 Off |                    0 |
| N/A   25C    P8              14W /  70W |      2MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+

/app/env/lib/python3.10/site-packages/torch/cuda/__init__.py:129: UserWarning: CUDA initialization: Unexpected error from cudaGetDeviceCount(). Did you run some cuda functions before calling NumCudaDevices() that might have already set an error? Error 304: OS call failed or operation not supported on this OS (Triggered internally at /opt/conda/conda-bld/pytorch_1728929558238/work/c10/cuda/CUDAFunctions.cpp:108.)
  return torch._C._cuda_getDeviceCount() > 0

Using CPU

**Comment by abhishekkrthakur** [comment]
did you factory rebuild the competiton space after setting the new hardware? 🤔

**Comment by lennart-maack** [comment]
Hi Abhishek. thanks for the quick reply!

I did factory rebuild the competition space. However, I have the same problems as described above.

**Comment by abhishekkrthakur** [comment]
please share the competiton config and space id

**Comment by abhishekkrthakur** [comment]
this looks fine on first glance. torch, torchvision and torchaudio are already installed. so, im suspecting having them present in requirements.txt might be messing up with the installation. could you try removing them from requirements and try again?

**Comment by lennart-maack** [comment]
I tried two different requirments.txt files in the competition space:

1:
--extra-index-url https://download.pytorch.org/whl/cu113
pycocotools
pandas
timm
pillow
transformers
tqdm

2:
- I had not requirements.txt at all

I did a factory rebuild, submitted another script.py and get the same log as above again (`torch.cuda.is_available()`) returns false again..

**Comment by abhishekkrthakur** [comment]
okay. ill investigate further and come back to you asap (in a few hours). apologies for the inconvenience.

**Comment by abhishekkrthakur** [comment]
fixed by #48

**Comment by lennart-maack** [comment]
I can confirm the fix. Thank you for the quick help!
